sklearn/model_selection/_split.py

```diff
@@ -799,21 +832,32 @@ def split(self, X, y=None, groups=None):
         n_samples = _num_samples(X)
         n_splits = self.n_splits
         n_folds = n_splits + 1
+        gap_size = self.gap_size
+        test_size = self.test_size if self.test_size else n_samples // n_folds
```

Comment:
perhaps this should be `if self.test_size is not None` so that 0 and None are not treated the same.

Reply:
perhaps the else should be `(n_samples - gap) // n_folds` so that train_size remains > gap

Reply:
I've updated the `self.test_size is not None`. I'm not convinced that the train_size necessarily needs to be constrained to be greater than the gap, should that be a decision left to the user?

Reply:
My opinion is I like the idea of having sensible defaults and as much control for the user as possible:

- Add option `walk_fw=False` ; if set to `True`: disregard `n_splits` and walk forward with specified `min_train_size`,  `test_size` --> keep the functionality of `n_splits` as it is - add the previous `test_size` to next train_set
- Allow user to specify `min_train_size` smaller than `gap`, if they so desire (it's a parameter to be explored imo). I agree with you though that a reasonable default is `min_train_size == gap if gap > 0` 
- Allow user to specify a `min_train_size` both smaller and larger than `test_size` but keep default >= 'test_size'
- `max_train_size='fixed'` --> I'm missing something, what is the use of a max_train_size when doing walk forward while growing the train_set with the previous test_set
- What is the idea behind test set overlap from caret for walk forward in our case? We would have overlap if we walk forward for 1 time unit and our test_size spans more than 1 time unit. That is the idea? I may be missing something. On that note also consider the sliding window, we have overlap here for certain if we have a horizon or `test_size>1`:

What I would love to see added from caret as well is the idea of a sliding window. Please consider including this as well. Caret calls it `fixedWindow` though. The idea is to include only examples inside a fixed time window and then walk forward and move the window along --> could be implemented along the lines of  `if n_splits=='window' and min_train_size>1 and test_size>1`. Here, `test_size` is the prediction horizon of the window, `min_train_size` is the train_size for the window each fold that gets moved up by 1 each time.